### 문제 1.
mtcars 데이터셋의 mpg열 데이터의 평균이 20과 같다고 할 수 있는지 검정하시오. (유의수준 5%)

In [7]:
import pandas as pd
import numpy as np
df = pd.read_csv('mtcars.csv')
# print(df.head())

# 1. 가설설정
# H0 : mpg열 데이터의 평균은 20과 같다 (mu = 20)
# H1 : mpg열 데이터의 평균은 20과 같지 않다 (mu != 20)
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 정규성 검정
# H0 : 정규분포를 따른다
# H1 : 정규분포를 따르지 않는다
import scipy.stats as stats
from scipy.stats import shapiro
statistic, pvalue = stats.shapiro(df['mpg'])
# print(round(statistic, 4), round(pvalue, 4))
# 귀무가설 채택
# 4.1 검정 실시 (정규성O)
statistic, pvalue = stats.ttest_1samp(df['mpg'], popmean=20, alternative='two-sided')
print(round(statistic, 4), round(pvalue, 4))
# 4.2 (정규성X)
statistic2, pvalue2 = stats.wilcoxon(df['mpg']-20, alternative='two-sided')
print(round(statistic2, 4), round(pvalue2, 4))
# 5. 귀무가설 기각여부 결정
# p-value가 0.05보다 크므로 귀무가설 채택
# 즉, mpg열 데이터의 평균은 20과 같다고 할 수 있다.
# 답 : 채택

0.0851 0.9328
249.0 0.7891


### 문제 2.
mtcars 데이터셋의 mpg열 데이터의 평균이 17보다 크다고 할 수 있는지 검정하시오. (유의수준 5%)

In [13]:
import pandas as pd
import numpy as np
df = pd.read_csv('mtcars.csv')
#print(df.head())
# 1. 가설설정
# H0 : mpg열 데이터의 평균은 17보다 작거나 같다
# H1 : mpg열 데이터의 평균은 17보다 크다
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 정규성 검정
# H0 : 정규분포를 따른다.
# H1 : 정규분포를 따르지 않는다.
import scipy.stats as stats
from scipy.stats import shapiro
statistic, pvalue = stats.shapiro(df['mpg'])
#print(round(statistic, 4), round(pvalue, 4))
# 정규분포를 따른다.
# 4-1. 검정실시 (정규성O)
statistic, pvalue = stats.ttest_1samp(df['mpg'], popmean=17, alternative='greater')
print(round(statistic, 4), round(pvalue, 4))
# 4-2. (정규성X)
statistic2, pvalue2 = stats.wilcoxon(df['mpg']-17, alternative='greater')
print(round(statistic2, 4), round(pvalue2, 4))
# 5. 귀무가설 기각여부 결정
# p-value값이 0.05보다 작으므로 귀무가설 기각. 대립가설 채택
# 즉, mpg열 데이터의 평균은 17보다 크다고 할 수 있다.
# 답 : 기각

2.9008 0.0034
395.5 0.0066


### 문제 3. 
mtcars 데이터셋의 mpg열 데이터의 평균이 17보다 작다고 할 수 있는지 검정하시오. (유의수준 5%)

In [18]:
import pandas as pd
import numpy as np
df = pd.read_csv('mtcars.csv')
#print(df.head())
# 1. 가설설정
# H0 : mpg열 데이터의 평균이 17보다 크거나 같다.
# H1 : mpg열 데이터의 평균이 17보다 작다.
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 정규성 검정
# H0 : 정규분포를 따른다.
# H1 : 정규분포를 따르지 않는다.
import scipy.stats as stats
from scipy.stats import shapiro
statistic, pvalue = stats.shapiro(df['mpg'])
#print(round(statistic, 4), round(pvalue, 4))
# 정규분포를 따른다.
# 4-1. 검정실시 (정규성O)
statistic, pvalue = stats.ttest_1samp(df['mpg'], popmean=17, alternative='less')
print(round(statistic, 4), round(pvalue, 4))
statistic2, pvalue2 = stats.wilcoxon(df['mpg']-17, alternative='less')
print(round(statistic2, 4), round(pvalue2, 4))
# 5. 귀무가설 기각여부 결정
# p-value 값이 0.05보다 크므로 귀무가설 채택
# 즉, mpg열 데이터의 평균이 17보다 크거나 같다고 할 수 있다.
# 답 : 채택

2.9008 0.9966
395.5 0.9938


### 문제 4.
다음은 혈압약을 먹은 전,후의 혈압 데이터이다.
혈압약을 먹기 전, 후의 차이가 있는지 쌍체 t검정을 실시하시오. (유의수준 5%)
before : 혈압약을 먹기 전 혈압, after : 혈압약을 먹은 후의 혈압

In [28]:
import pandas as pd
import numpy as np
#	데이터	만들기
df	=	pd.DataFrame(	{
	'before':	[120,	135,	122,	124,	135,	122,	145,	160,	155,	142,	144,	135,	167],
	'after'	:	[110,	132,	123,	119,	123,	115,	140,	162,	142,	138,	135,	142,	160]	})
#print(df.head(3))
# 1. 가설설정
# H0 : 혈압약을 먹기 전, 후의 차이가 없다 after - before = 0
# H1 : 혈압약을 먹기 전, 후의 차이가 있다 after - before != 0
# 2. 유의수준 확인 : 유의수준 5%
# 3. 정규성 확인
# H0 : 차이값이 정규분포를 따른다.
# H1 : 차이값이 정규분포를 따르지 않는다.
import scipy.stats as stats
from scipy.stats import shapiro
statistic, pvalue = stats.shapiro(df['after']-df['before'])
#print(round(statistic, 4), round(pvalue, 4))
# 정규분포를 따른다
# 4-1. 검정실시 (정규성O)
statistic, pvalue = stats.ttest_rel(df['after'], df['before'], alternative='two-sided')
print(round(statistic, 4), round(pvalue, 4))
# 4-2. (정규성X)
statistic2, pvalue2 = stats.wilcoxon(df['after']-df['before'], alternative='two-sided')
print(round(statistic2, 4), round(pvalue2, 4))
# 5. 귀무가설 기각여부 결정
# p-value 가 0.05보다 작으므로 귀무가설 기각
# 즉, 혈압약을 먹기 전, 후의 차이가 있다
# 답 : 기각

-3.1382 0.0086
11.0 0.0134


### 문제 5.
다음은 혈압약을 먹은 전, 후의 혈압 데이터이다.
혈압약을 먹은 후 혈압이 감소했는지 확인하기 위해 쌍체 t검정을 실시하시오. (유의수준 5%)

In [33]:
 #	데이터	만들기
df	=	pd.DataFrame(	{
	'before':	[120,	135,	122,	124,	135,	122,	145,	160,	155,	142,	144,	135,	167],
	'after'	:	[110,	132,	123,	119,	123,	115,	140,	162,	142,	138,	135,	142,	160]	})
#print(df.head(3))
# 1. 가설설정
# H0 : 혈압약을 먹은후 혈압이 같거나 증가했다. after - before >= 0
# H1 : 혈압약을 먹은후 혈압이 감소했다. after - before < 0
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 정규성 검정
# H0 : 차이값이 정규분포를 따른다
# H1 : 차이값이 정규분포를 따르지 않는다
import scipy.stats as stats
from scipy.stats import shapiro
statistic, pvalue = stats.shapiro(df['after']-df['before'])
#print(round(statistic, 4), round(pvalue, 4))
# 정규분포를 따른다.
# 4-1. 검정실시 (정규성O)
statistic, pvalue = stats.ttest_rel(df['after'], df['before'], alternative='less')
print(round(statistic, 4), round(pvalue, 4))
# 4-2. (정규성X)
statistic2, pvalue2 = stats.wilcoxon(df['after']-df['before'], alternative='less')
print(round(statistic2, 4), round(pvalue2, 4))
# 5. 귀무가설 기각여부 결정
# p-value가 0.05보다 작으므로 귀무가설 기각
# 즉, 혈압약을 먹은 후 혈압이 감소했다.
# 답 : 기각

-3.1382 0.0043
11.0 0.0067


### 문제 6.
다음은 A그룹과 B그룹 인원의 혈압 데이터이다.
두 그룹의 혈압평균이 다르다고 할 수 있는지 독립표본 t검정을 실시하시오. (유의수준 5%)

In [40]:
#	데이터	만들기
df = pd.DataFrame({
				'A': [120,	135,	122,	124,	135,	122,	145,	160,	155,	142,	144,	135,	167],
				'B'	: [110,	132,	123,	119,	123,	115,	140,	162,	142,	138,	135,	142,	160]})
#print(df.head(3))
# 1. 가설설정
# H0 : 두 그룹의 혈압평균이 같다 A = B
# H1 : 두 그룹의 혈압평균이 다르다 A != B
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 정규성 검정
# H0 : 두 그룹 모두 정규분포를 따른다
# H1 : not H0
import scipy.stats as stats
from scipy.stats import shapiro
statisticA, pvalueA = stats.shapiro(df['A'])
statisticB, pvalueB = stats.shapiro(df['B'])
print(round(statisticA, 4), round(pvalueA, 4))
print(round(statisticB, 4), round(pvalueB, 4))
# 두 그룹 모두 정규분포를 따른다
# 4. 등분산 검정
# H0 : 등분산한다
# H1 : 등분산하지 않는다
statistic, pvalue = stats.bartlett(df['A'], df['B'])
print(round(statistic, 4), round(pvalue, 4))
# 등분산한다
# 5-1. 검정실시 (정규성O, 등분산O)
statistic, pvalue = stats.ttest_ind(df['A'], df['B'], equal_var=True, alternative='two-sided')
print(round(statistic, 4), round(pvalue, 4))
# 5-2. (정규성X) 순위합검정
statistic, pvalue = stats.ranksums(df['A'], df['B'], alternative='two-sided')
print(round(statistic, 4), round(pvalue, 4))
# 6. 귀무가설 기각여부 결정
# p-value 가 0.05보다 크므로 귀무가설 채택
# 즉, A와 B의 혈압평균은 같다고 할 수 있다
# 답 : 채택

0.9314 0.3558
0.9498 0.5956
0.0279 0.8673
0.8192 0.4207
0.8462 0.3975


### 문제 7.
다음은 A그룹과 B그룹 인원의 혈압 데이터이다. A그룹의 혈압 평균이 B그룹보다 크다고 할 수 있는지
독립표본 t검정을 실시하시오. (유의수준 5%)

In [45]:
#	데이터	만들기
df = pd.DataFrame({
				'A': [120,	135,	122,	124,	135,	122,	145,	160,	155,	142,	144,	135,	167],
				'B'	: [110,	132,	123,	119,	123,	115,	140,	162,	142,	138,	135,	142,	160]})
#print(df.head(3))
# 1. 가설설정
# H0 : A그룹의 혈압 평균이 B그룹보다 작거나 같다. A <= B
# H1 : A그룹의 혈압 평균이 B그룹보다 크다. A > B
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 정규성 검정
# H0 : 정규분포를 따른다
# H1 : 정규분포를 따르지 않는다
import scipy.stats as stats
from scipy.stats import shapiro
statisticA, pvalueA = stats.shapiro(df['A'])
statisticB, pvalueB = stats.shapiro(df['B'])
print(round(statisticA, 4), round(pvalueA, 4))
print(round(statisticB, 4), round(pvalueB, 4))
# 정규분포를 따른다
# 4. 등분산 검정
# H0 : 등분산 한다
# H1 : 등분산 하지 않는다
statistic, pvalue = stats.bartlett(df['A'], df['B'])
print(round(statistic, 4), round(pvalue, 4))
# 등분산 한다
# 5-1. 검정실시 (정규성O)
statistic, pvalue = stats.ttest_ind(df['A'], df['B'], equal_var=True, alternative='greater')
print(round(statistic, 4), round(pvalue, 4))
# 5-2. (정규성X)
statistic, pvalue = stats.ranksums(df['A'], df['B'], alternative='greater')
print(round(statistic, 4), round(pvalue, 4))
# 6. 귀무가설 기각여부 결정
# p-value가 0.05보다 크므로 귀무가설 채택
# 즉, A 혈압 평균이 B 혈압 평균보다 크다고 할 수 없다. 작거나 같다
# 답 : 채택

0.9314 0.3558
0.9498 0.5956
0.0279 0.8673
0.8192 0.2104
0.8462 0.1987


### 문제 8.
다음은 A, B, C 그룹 인원 성적 데이터이다.
세 그룹의 성적 평균이 같다고 할 수 있는지 ANOVA 분석을 실시하시오. (유의수준 5%)

In [51]:
 #	데이터	만들기
df = pd.DataFrame(	{
	'A': [120,	135,	122,	124,	135,	122,	145,	160,	155,	142,	144,	135,	167],
	'B'	: [110,	132,	123,	119,	123,	115,	140,	162,	142,	138,	135,	142,	160],
	'C'	: [130,	120,	115,	122,	133,	144,	122,	120,	110,	134,	125,	122,	122]})
#print(df.head(3))
# 1. 가설설정
# H0 : 세 그룹의 성적 평균이 같다. A = B = C
# H1 : 적어도 하나는 같지 않다. not H0
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 정규성 검정
import scipy.stats as stats
from scipy.stats import shapiro
statisticA, pvalueA = stats.shapiro(df['A'])
statisticB, pvalueB = stats.shapiro(df['B'])
statisticC, pvalueC = stats.shapiro(df['C'])
print(round(statisticA, 4), round(pvalueA, 4))
print(round(statisticB, 4), round(pvalueB, 4))
print(round(statisticC, 4), round(pvalueC, 4))
# 정규분포를 따른다.
# 4. 등분산 검정
statistic, pvalue = stats.bartlett(df['A'], df['B'], df['C'])
print(round(statistic, 4), round(pvalue, 4))
# 등분산 한다.
# 5-1. 검정실시 (정규성O, 등분산O)
statistic, pvalue = stats.f_oneway(df['A'], df['B'], df['C'])
print(round(statistic, 4), round(pvalue, 4))
# (정규성O, 등분산X) welch_anova
# 5-2. (정규성X) 크루스칼 왈리스 검정
statistic, pvalue = stats.kruskal(df['A'], df['B'], df['C'])
print(round(statistic, 4), round(pvalue, 4))
# 6. 귀무가설 기각여부 결정
# p-value 가 0.05보다 작으므로 귀무가설 기각
# 즉, 적어도 하나는 같지 않다
# 답 : 기각

0.9314 0.3558
0.9498 0.5956
0.9397 0.4527
4.2222 0.1211
3.6971 0.0346
6.897 0.0318


데이터 형태가 다를 경우

In [53]:
#	데이터	만들기
df2	= pd.DataFrame(	{
 '항목': ['A','A','A','A','A','A','A','A','A','A','A','A','A',
         'B','B','B','B','B','B','B','B','B','B','B','B','B',
         'C','C','C','C','C','C','C','C','C','C','C','C','C',],
 'value': [120,	135,	122,	124,	135,	122,	145,	160,	155,	142,	144,	135,	167,
 110,	132,	123,	119,	123,	115,	140,	162,	142,	138,	135,	142,	160,
 130,	120,	115,	122,	133,	144,	122,	120,	110,	134,	125,	122,	122]
 })
#print(df2.head(3))
a = df2[df2['항목']=='A']['value']
b = df2[df2['항목']=='B']['value']
c = df2[df2['항목']=='C']['value']
print(stats.f_oneway(a,b,c))

F_onewayResult(statistic=3.6970511827172867, pvalue=0.0346491232068625)


### 문제 9.
랜덤 박스에 상품 A,B,C,D 가 들어있다.
다음은 랜덤박스에서 100번 상품을 꺼냈을 때의 상품 데이터라고 할 때
상품이 동일한 비율로 들어있다고 할 수 있는지 검정해보시오.

In [58]:
 #	데이터	생성
row1 = [30,	20,	15,	35]
df = pd.DataFrame([row1], columns=['A','B','C',	'D'])
#df
# 1. 가설설정
# H0 : 상품이 동일한 비율로 들어있다
# H1 : 상품이 동일한 비율로 들어있지 않다
# 2. 유의수준 확인 (유의수준 5%)
# 3. 검정실시
import scipy.stats as stats
from scipy.stats import chisquare
statistic, pvalue = stats.chisquare(f_obs=df.iloc[0], f_exp=[25,25,25,25])
print(round(statistic, 4), round(pvalue, 4))
# 4. 귀무가설 기각여부 결정
# p-value 가 0.05보다 작으므로 귀무가설 기각
# 즉, 동일한 비율로 들어있지 않다.
# 답 : 기각

10.0 0.0186


### 문제 10.
랜덤 박스에 상품 A,B,C가 들어있다.
다음은 랜덤박스에서 150번 상품을 꺼냈을 때의 상품 데이터라고 할 때
상품별로 A 30%, B 15%, C 55% 비율로 들어있다고 할 수 있는지 검정해보시오. 

In [61]:
 #	데이터	생성
row1 = [50,25,75]
df = pd.DataFrame([row1], columns=['A','B','C'])
#df
# 1. 가설설정
# H0 : 상품별로 30, 15, 55 비율로 들어있다
# H1 : 상품별로 30, 15, 55 비율로 들어있지 않다
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 검정실시
import scipy.stats as stats
from scipy.stats import chisquare
statistic, pvalue = chisquare(f_obs=df.iloc[0], f_exp=[150*0.3, 150*0.15, 150*0.55])
print(round(statistic, 4), round(pvalue, 4))
# 4. 귀무가설 기각여부 결정
# p-value 가 0.05보다 크므로 귀무가설 채택
# 즉, 30,15,55 비율로 들어있다고 할 수 있다
# 답 : 채택

1.5152 0.4688


### 문제 11.
연령대에 따라 먹는 아이스크림의 차이가 있는지 독립성 검정을 실시하시오.

In [73]:
 #	데이터	생성
row1, row2 = [200,	190, 250], [220, 250, 300]
df = pd.DataFrame([row1,	row2],	columns=['딸기','초코','바닐라'],	index=['10대',	'20대'])
#print(df)
# 1. 가설설정
# H0 : 연령대에 따라 먹는 아이스크림의 차이가 없다. 두 범주형 변수는 서로 독립이다.
# H1 : 연령대에 따라 먹는 아이스크림의 차이가 있다. 두 범주형 변수는 서로 독립이 아니다.
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 검정실시
import scipy.stats as stats
from scipy.stats import chi2_contingency
statistic, pvalue, dof, expected = chi2_contingency(df)
print(statistic)
print(pvalue)
print(dof)
print(np.round(expected, 2))
print(chi2_contingency([df.iloc[0], df.iloc[1]]))
# 4. 귀무가설 기각여부 결정
# p-value 값이 0.05보다 크므로 귀무가설 채택
# 즉, 두 범주형 변수는 서로 독립이다.
# 답 : 채택

1.708360126075226
0.4256320394874311
2
[[190.64 199.72 249.65]
 [229.36 240.28 300.35]]
Chi2ContingencyResult(statistic=1.708360126075226, pvalue=0.4256320394874311, dof=2, expected_freq=array([[190.63829787, 199.71631206, 249.64539007],
       [229.36170213, 240.28368794, 300.35460993]]))


데이터 형태가 다를 경우

In [76]:
 #	★	tip	:	pd.crosstab()	사용방법
#	(Case1)	만약	데이터가	아래와	같이	주어진다면?
df	=	pd.DataFrame({
 '아이스크림'	:	['딸기','초코','바닐라','딸기','초코','바닐라'],
 '연령'	:	['10대','10대','10대','20대','20대','20대'],
 '인원'	:	[200,190,250,220,250,300]
 })
#df
table = pd.crosstab(index= df['연령'], columns= df['아이스크림'], values= df['인원'], aggfunc='sum')
#print(table)
statistic, pvalue, dof, expected = chi2_contingency([table.iloc[0], table.iloc[1]])
print(statistic)
print(pvalue)
print(dof)
print(np.round(expected, 2))

1.708360126075226
0.4256320394874311
2
[[190.64 249.65 199.72]
 [229.36 300.35 240.28]]


In [80]:
# (Case 2) 만약 데이터가 아래와 같이 주어진다면?
# (이해를 위한 참고용, 빈도수 카운팅)
df	=	pd.DataFrame({
 '아이스크림'	:	['딸기','초코','바닐라','딸기','초코','바닐라'],
 '연령'	:	['10대','10대','10대','20대','20대','20대'],})
#df
table = pd.crosstab(df['연령'], df['아이스크림'])
table

아이스크림  딸기  바닐라  초코
연령                
10대     1    1   1
20대     1    1   1

### 문제 12.
타이타닉 데이터에서 성별(sex)과 생존여부(survived) 변수간 독립성 검정을 실시하시오.

In [87]:
import seaborn as sns
df = sns.load_dataset('titanic')
#df.info()
#print(df['sex'].value_counts())
#print(df['survived'].value_counts())
table = pd.crosstab(index=df['sex'], columns=df['survived'])
#print(table)
# 1. 가설설정
# H0 : 성별과 생존여부는 관련이 없다 (독립이다)
# H1 : 성별과 생존여부는 관련이 있다 (독립이 아니다)
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 검정실시
import scipy.stats as stats
from scipy.stats import chi2_contingency
statistic, pvalue, dof, expected = chi2_contingency([table.iloc[0], table.iloc[1]])
print(statistic)
print(pvalue)
print(dof)
print(np.round(expected, 2))
# 4. 귀무가설 기각여부 결정
# p-value 가 0.05보다 작으므로 귀무가설 기각
# 즉, 성별과 생존여부는 관련이 있다 (독립이 아니다)
# 답 : 기각

260.71702016732104
1.1973570627755645e-58
1
[[193.47 120.53]
 [355.53 221.47]]


데이터 변경

In [90]:
# 임의 데이터 생성
sex, survived = [160,160], [250,220]
table = pd.DataFrame([sex,survived], columns=['0','1'], index=['female','male'])
#print(table)
print(chi2_contingency(table))
# 귀무가설 채택
# 독립이다

Chi2ContingencyResult(statistic=0.6541895872879862, pvalue=0.41861876333789727, dof=1, expected_freq=array([[166.07594937, 153.92405063],
       [243.92405063, 226.07594937]]))


### 문제 1.
- 다중회귀분석을 실시하고 각 물음에 답하시오
- 독립변수 : age, sex, bmi
- 종속변수 : target

- 1-1) Rsq(결정계수) 값을 구하시오.
- 1-2) 회귀계수를 모두 구하시오.
- 1-3) 회귀식의 p-value 값을 구하시오.

In [106]:
###############		실기환경	복사	영역		###############
 #	데이터	불러오기
import	pandas	as	pd
import	numpy	as	np
#	실기	시험	데이터셋으로	셋팅하기	(수정금지)
from	sklearn.datasets	import	load_diabetes
 #	diabetes	데이터셋	로드
diabetes	=	load_diabetes()
x	=	pd.DataFrame(diabetes.data,	columns=diabetes.feature_names)
y	=	pd.DataFrame(diabetes.target)
y.columns	=	['target']
 ###############		실기환경	복사	영역		###############
x = x[['age','sex','bmi']]

# 1. sklearn으로 풀기
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(x,y)
# 결정계수
print(np.round(model.score(x,y), 2))
# 회귀계수
print(np.round(model.coef_, 4))
print(np.round(model.coef_[0,0], 4))
print(np.round(model.coef_[0,1], 4))
print(np.round(model.coef_[0,2], 4))
# 절편
print(np.round(model.intercept_, 4))
# 회귀식
# y = 152.1335 + 138.9039age - 36.1353sex + 926.912bmi

0.35
[[138.9039 -36.1353 926.912 ]]
138.9039
-36.1353
926.912
[152.1335]


In [108]:
###############		실기환경	복사	영역		###############
 #	데이터	불러오기
import	pandas	as	pd
import	numpy	as	np
#	실기	시험	데이터셋으로	셋팅하기	(수정금지)
from	sklearn.datasets	import	load_diabetes
 #	diabetes	데이터셋	로드
diabetes	=	load_diabetes()
x	=	pd.DataFrame(diabetes.data,	columns=diabetes.feature_names)
y	=	pd.DataFrame(diabetes.target)
y.columns	=	['target']
 ###############		실기환경	복사	영역		###############
x = x[['age','sex','bmi']]

# 2. statsmodel로 풀이
import statsmodels.api as sm

x = sm.add_constant(x)

model = sm.OLS(y, x).fit()
summary = model.summary()
print(summary)
# p-value : 7.77e-41

                            OLS Regression Results                            
Dep. Variable:                 target   R-squared:                       0.351
Model:                            OLS   Adj. R-squared:                  0.346
Method:                 Least Squares   F-statistic:                     78.94
Date:                Wed, 19 Jun 2024   Prob (F-statistic):           7.77e-41
Time:                        22:43:16   Log-Likelihood:                -2451.6
No. Observations:                 442   AIC:                             4911.
Df Residuals:                     438   BIC:                             4928.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.1335      2.964     51.321      0.0

### 문제 2.
- 두 변수(bmi, target)에 대해 피어슨 상관분석을 실시하고 아래 물음에 답하시오.
- 2-1) 귀무가설과 대립가설을 설정하시오.
- 2-2) 상관계수 값을 구하시오.
- 2-3) 검정통계량과 p-value 값을 구하시오.
- 2-4) 귀무가설 채택/기각 여부를 결정하시오.

In [116]:
###############		실기환경	복사	영역		###############
 #	데이터	불러오기
import	pandas	as	pd
import	numpy	as	np
#	실기	시험	데이터셋으로	셋팅하기	(수정금지)
from	sklearn.datasets	import	load_diabetes
 #	diabetes	데이터셋	로드
diabetes	=	load_diabetes()
x	=	pd.DataFrame(diabetes.data,	columns=diabetes.feature_names)
y	=	pd.DataFrame(diabetes.target)
y.columns	=	['target']
 ###############		실기환경	복사	영역		###############
x = x['bmi']
y = y['target']

# 1. 가설설정
# H0 : 두 변수는 선형관계가 없다. p = 0
# H1 : 두 변수는 선형관계가 있다. p != 0
# 2. 유의수준 확인 : 유의수준 5%로 확인
# 3. 검정실시
import scipy.stats as stats
from scipy.stats import pearsonr

r, pvalue = pearsonr(x, y)
print(r)
print(pvalue)
n = len(x)

statistic = r * np.sqrt((n-2)/(1-r**2))
print(statistic)
# 4. 귀무가설 기각여부 결정
# p-value 가 0.05보다 작으므로 귀무가설 기각
# 즉, 선형관계가 있다.
# 답 : 기각

0.5864501344746887
3.466006445167154e-42
15.187289570365309


### 문제 3.
- 로지스틱회귀분석을 실시하고 각 물음에 답하시오.
- 독립변수 : sex, sibsp, fare
- 종속변수 : survived
- 3-1) 회귀계수를 모두 구하시오.
- 3-2) 로지스틱 회귀모형에서 sibsp 변수가 한 단위 증가할 때 생존할 오즈가 몇 배 증가하는지 반올림하여 소수점 셋째 자리까지 구하시오.
- 3-3) 로지스틱 회귀모형에서 여성일 경우 남성에 비해 오즈가 몇 배 증가하는지 반올림하여 소수점 셋째 자리까지 구하시오.

In [127]:
###############  복사 영역  ###############
 # 데이터 불러오기
import pandas as pd
import numpy as np
 # Seaborn의 내장 타이타닉 데이터셋을 불러옵니다.
import seaborn as sns
df = sns.load_dataset('titanic')
 # 분석 데이터 설정
df = df[ ['survived','sex','sibsp','fare'] ] # sex:성별, sibsp:탑승한 부모 및 자녀
# 성별을 map 함수를 활용해서 각각 1과 0에 할당한다.(여성을 1, 남성을 0)
 # (실제 시험의 지시 조건에 따를 것)
df['sex'] = df['sex'].map({'female': 1,'male': 0 })
 ###############  복사 영역  ##############
x = df.drop(['survived'], axis=1)
y = df['survived']
# print(x)
# print(y)

# 1. sklearn으로 구하기
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(penalty=None)
model.fit(x, y)
# 회귀계수
print(np.round(model.coef_, 4))
print(np.round(model.coef_[0,0], 4))
print(np.round(model.coef_[0,1], 4))
print(np.round(model.coef_[0,2], 4))
# 절편
print(np.round(model.intercept_, 4))
# 1번 답 회귀식 y = -1.6964 + 2.5668sex - 0.4017sibsp + 0.0138fare
# 오즈 P / 1-P = exp^(f(x))
# 로짓 f(x)
# 2번 답 0.669배  증가
print(round(np.exp(model.coef_[0,1]), 3))
# 3번 답 13.024배 증가
print(round(np.exp(model.coef_[0,0]), 3))

[[ 2.5668 -0.4017  0.0138]]
2.5668
-0.4017
0.0138
[-1.6964]
0.669
13.024


In [131]:
###############  복사 영역  ###############
 # 데이터 불러오기
import pandas as pd
import numpy as np
 # Seaborn의 내장 타이타닉 데이터셋을 불러옵니다.
import seaborn as sns
df = sns.load_dataset('titanic')
 # 분석 데이터 설정
df = df[ ['survived','sex','sibsp','fare'] ] # sex:성별, sibsp:탑승한 부모 및 자녀
# 성별을 map 함수를 활용해서 각각 1과 0에 할당한다.(여성을 1, 남성을 0)
 # (실제 시험의 지시 조건에 따를 것)
df['sex'] = df['sex'].map({'female': 1,'male': 0 })
 ###############  복사 영역  ##############
x = df.drop(['survived'], axis=1)
y = df['survived']

# 2. statsmodels 로 구하기
import statsmodels.api as sm

x = sm.add_constant(x)

model = sm.Logit(y, x).fit()

summary = model.summary()
print(summary)
print(round(np.exp(-0.4017), 3))
print(round(np.exp(2.5668), 3))

Optimization terminated successfully.
         Current function value: 0.483846
         Iterations 6
                           Logit Regression Results                           
Dep. Variable:               survived   No. Observations:                  891
Model:                          Logit   Df Residuals:                      887
Method:                           MLE   Df Model:                            3
Date:                Wed, 19 Jun 2024   Pseudo R-squ.:                  0.2734
Time:                        23:15:23   Log-Likelihood:                -431.11
converged:                       True   LL-Null:                       -593.33
Covariance Type:            nonrobust   LLR p-value:                 5.094e-70
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.6964      0.129    -13.134      0.000      -1.950      -1.443
sex            2.5668      0.

### 체험환경 문제

In [ ]:
# 1. gender와 survived 변수 간의 독립성 검정을 실시하였을때, 카이제곱 통계량은? (반올림하여 소수 셋째 자리까지 계산)
import pandas as pd
import numpy as np
df = pd.read_csv("data/Titanic.csv")

# 사용자 코딩
#print(df)
# 1. 독립성검정
x = df['Gender']
y = df['Survived']

#print(x)
#print(y)
table = pd.crosstab(index=x, columns=y)
#print(table)
# 1. 가설설정
# H0 : 두 변수는 관련이 없다. 독립이다
# H1 : 두 변수는 관련이 있다. 독립이 아니다
# 2. 유의수준 설정 5%
# 3. 검정실시
import scipy.stats as stats
from scipy.stats import chi2_contingency
statistic, pvalue, dof, expected = chi2_contingency([table.iloc[0], table.iloc[1]])
print(round(statistic, 3))
print(round(pvalue, 3))
print(dof)
print(np.round(expected, 3))

In [ ]:
# 2. gender, sibsp, parch, fare를 독립변수로 사용하여 로지스틱 회귀모형을 실시하였을 때, parch 변수의 계수값은?
# (반올림하여 소수 셋째 자리까지 계산)
import pandas as pd
import numpy as np
df = pd.read_csv("data/Titanic.csv")
df['Gender'] = df['Gender'].map({'female':1, 'male':0})

# 사용자 코딩
x = df[['Gender','SibSp','Parch','Fare']]
y = df['Survived']
# print(x)
# print(y)
# 1. sklearn 활용
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(penalty='none')
model.fit(x, y)
# 회귀계수
print(np.round(model.coef_, 3))
print(np.round(model.intercept_, 3))
# y = -1.696 + 2.642gender - 0.354sibsp - 0.201parch + 0.015fare

# 2. statsmodels 활용
import statsmodels.api as sm

x = sm.add_constant(x)

model = sm.Logit(y, x).fit()

summary = model.summary()
print(summary)

In [ ]:
# 3. 위 2번 문제에서 추정된 로지스틱 회귀모형에서 SibSp 변수가 한 단위 증가할 때 생존할 오즈비(Odds ratio) 값은?
# (반올림하여 소수 셋째 자리까지 계산)
# 1. sklearn
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(penalty='none')
model.fit(x, y)
print(np.round(model.coef_, 4))
print(np.round(model.coef_[0,1], 4))
# exp^(b2)
print(round(np.exp(model.coef_[0,1]), 3))

# 2. statsmodels
import statsmodels.api as sm

x = sm.add_constant(x)

model = sm.Logit(y, x).fit()
summary = model.summary()
#print(summary)
print(round(np.exp(-0.3539), 3))

### 7회 기출
- 문제 3-1.
- survived를 종속변수(Y), sex,sipsp,fare 변수를 독립변수(X)로 분석했을 때 아래 질문에 답하시오.
- (단, 총 891개의 데이터에서 초기 500개 데이터로 학습, 391개 데이터로 평가 하시오.)
- (단, 모든 답은 반올림하여 소수점 둘째자리까지 구하시오.)
- 1. 초기 500개 데이터로 분석 시 sibsp변수의 odds ratio는?
  2. 초기 500개 데이터로 분석시 residual deviance는?
  3. 나머지 391개 데이터 적용시 오분류율은?

In [21]:
# 데이터 생성
import pandas as pd
import numpy as np
import seaborn as sns
df = sns.load_dataset('titanic')

df = df[['survived','sex','sibsp','fare']]
# sex:성별, sibsp:탑승한 부모 및 자녀 수, fare:요금
df['sex'] = df['sex'].map({'female':1, 'male':0})

x_train = df.drop(columns=['survived']).loc[0:499]
x_test = df.drop(columns=['survived']).loc[500:]
y_train = df['survived'].loc[0:499]
y_test = df['survived'].loc[500:]

# print(x_train.shape)
# print(x_test.shape)
# print(y_train.shape)
# print(y_test.shape)
# print(x_train)
# print(y_train)

# 1. sklearn
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(penalty=None)
model.fit(x_train, y_train)

# print(np.round(model.coef_, 4))
# print(np.round(model.coef_[0,0], 4))
# print(np.round(model.coef_[0,1], 4))
# print(np.round(model.coef_[0,2], 4))
# print(np.round(model.intercept_, 4))

# exp^(b2)
#print(round(np.exp(model.coef_[0,1]), 2))
# 1. 0.68

from sklearn.metrics import log_loss

y_pred_prob = model.predict_proba(x_train)

logloss = log_loss(y_train, y_pred_prob, normalize=False)
print(logloss)
result = logloss*2
print(round(result, 2))
# 2. 477.07

y_result = model.predict(x_test)

from sklearn.metrics import accuracy_score
acc = accuracy_score(y_test, y_result)
print(acc)
print(round(1-acc, 2))
# 3. 0.22

238.5369401214462
477.07
0.782608695652174
0.22


- 문제 3-1.
- survived를 종속변수(Y), sex,sipsp,fare 변수를 독립변수(X)로 분석했을 때 아래 질문에 답하시오.
- (단, 총 891개의 데이터에서 초기 500개 데이터로 학습, 391개 데이터로 평가 하시오.)
- (단, 모든 답은 반올림하여 소수점 둘째자리까지 구하시오.)
- 1. 초기 500개 데이터로 분석 시 sibsp변수의 odds ratio는?
  2. 초기 500개 데이터로 분석시 residual deviance는?
  3. 나머지 391개 데이터 적용시 오분류율은?

In [28]:
# 데이터 생성
import pandas as pd
import numpy as np
import seaborn as sns
df = sns.load_dataset('titanic')

df = df[['survived','sex','sibsp','fare']]
# sex:성별, sibsp:탑승한 부모 및 자녀 수, fare:요금
df['sex'] = df['sex'].map({'female':1, 'male':0})

x_train = df.drop(columns=['survived']).loc[0:499]
x_test = df.drop(columns=['survived']).loc[500:]
y_train = df['survived'].loc[0:499]
y_test = df['survived'].loc[500:]

# 2. statsmodels 활용
import statsmodels.api as sm

x_train = sm.add_constant(x_train)

model = sm.Logit(y_train, x_train).fit()

summary = model.summary()
#print(summary)

#print(round(np.exp(-0.3799), 2))
# 1. 0.68

cross_entropy = -1 * -238.54
result = cross_entropy*2
#print(round(result, 2))
# 2. 477.08

x_test = sm.add_constant(x_test)

y_result_prob = model.predict(x_test)
y_result = np.where(y_result_prob>0.5, 1, 0)
#print(y_result)
from sklearn.metrics import accuracy_score
acc = accuracy_score(y_test, y_result)
print(acc)
print(round(1-acc, 2))
# 3. 0.22

Optimization terminated successfully.
         Current function value: 0.477074
         Iterations 6
0.782608695652174
0.22


In [32]:
# 데이터 생성
import pandas as pd
import numpy as np
import seaborn as sns
df = sns.load_dataset('titanic')

df = df[['survived','sex','sibsp','fare']]
# sex:성별, sibsp:탑승한 부모 및 자녀 수, fare:요금
df['sex'] = df['sex'].map({'female':1, 'male':0})

x_train = df.drop(columns=['survived']).loc[0:499]
x_test = df.drop(columns=['survived']).loc[500:]
y_train = df['survived'].loc[0:499]
y_test = df['survived'].loc[500:]

# GLM 모델
import statsmodels.api as sm

x_train = sm.add_constant(x_train)
model = sm.GLM(y_train, x_train, family=sm.families.Binomial()).fit()
summary = model.summary()
print(summary)

x_test = sm.add_constant(x_test)

y_result_prob = model.predict(x_test)
#print(y_result_prob)
y_result = np.where(y_result_prob>0.5, 1, 0)
from sklearn.metrics import accuracy_score
acc = accuracy_score(y_test, y_result)
print(acc)
print(round(1-acc, 2))

                 Generalized Linear Model Regression Results                  
Dep. Variable:               survived   No. Observations:                  500
Model:                            GLM   Df Residuals:                      496
Model Family:                Binomial   Df Model:                            3
Link Function:                  Logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -238.54
Date:                Thu, 20 Jun 2024   Deviance:                       477.07
Time:                        01:38:17   Pearson chi2:                     498.
No. Iterations:                     5   Pseudo R-squ. (CS):             0.3159
Covariance Type:            nonrobust                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.6815      0.174     -9.690      0.0

- 문제 3-2.
- 다음은 당뇨병 환자의 질병 진행정도 데이터셋이다.
- target을 종속변수(Y), 나머지를 독립변수(X)로 분석했을 때, 아래 질문에 답하시오.
- (단, 모든 답은 반올림하여 소수점 둘째자리까지 구하시오.)
- 1. target 변수와 가장 큰 상관관계를 갖는 변수의 상관계수를 구하시오.
  2. 다중선형회귀 모델링 후 결정계수(R2 score)를 구하시오.
  3. 문제 2에서 구한 회귀모델에서 p-value가 가장 큰 변수의 p-value 값을 구하시오.

In [45]:
###############		실기환경	복사	영역		###############
 #	데이터	불러오기
import	pandas	as	pd
import	numpy	as	np
#	실기	시험	데이터셋으로	셋팅하기	(수정금지)
from	sklearn.datasets	import	load_diabetes
 #	diabetes	데이터셋	로드
diabetes	=	load_diabetes()
x	=	pd.DataFrame(diabetes.data,	columns=diabetes.feature_names)
y	=	pd.DataFrame(diabetes.target)
y.columns	=	['target']
df = pd.concat([x,y], axis=1)
 ###############		실기환경	복사	영역		###############
# print(x)
# print(y)
#print(df)
#print(df.corr().abs().sort_values('target', ascending=False))
# 1. bmi 0.59
print(df.corr())

# 1. sklearn
x = df.drop(columns=['target'])
y = df['target']
# print(x)
# print(y)
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(x, y)
#print(round(model.score(x, y), 2))
# 2. 0.52

             age       sex       bmi        bp        s1        s2        s3  \
age     1.000000  0.173737  0.185085  0.335428  0.260061  0.219243 -0.075181   
sex     0.173737  1.000000  0.088161  0.241010  0.035277  0.142637 -0.379090   
bmi     0.185085  0.088161  1.000000  0.395411  0.249777  0.261170 -0.366811   
bp      0.335428  0.241010  0.395411  1.000000  0.242464  0.185548 -0.178762   
s1      0.260061  0.035277  0.249777  0.242464  1.000000  0.896663  0.051519   
s2      0.219243  0.142637  0.261170  0.185548  0.896663  1.000000 -0.196455   
s3     -0.075181 -0.379090 -0.366811 -0.178762  0.051519 -0.196455  1.000000   
s4      0.203841  0.332115  0.413807  0.257650  0.542207  0.659817 -0.738493   
s5      0.270774  0.149916  0.446157  0.393480  0.515503  0.318357 -0.398577   
s6      0.301731  0.208133  0.388680  0.390430  0.325717  0.290600 -0.273697   
target  0.187889  0.043062  0.586450  0.441482  0.212022  0.174054 -0.394789   

              s4        s5        s6   

LinearRegression()

In [44]:
###############		실기환경	복사	영역		###############
 #	데이터	불러오기
import	pandas	as	pd
import	numpy	as	np
#	실기	시험	데이터셋으로	셋팅하기	(수정금지)
from	sklearn.datasets	import	load_diabetes
 #	diabetes	데이터셋	로드
diabetes	=	load_diabetes()
x	=	pd.DataFrame(diabetes.data,	columns=diabetes.feature_names)
y	=	pd.DataFrame(diabetes.target)
y.columns	=	['target']
df = pd.concat([x,y], axis=1)
 ###############		실기환경	복사	영역		###############
x = df.drop(['target'], axis=1)
y = df['target']

# 2. statsmodels
import statsmodels.api as sm

x = sm.add_constant(x)

model = sm.OLS(y, x).fit()
summary = model.summary()
print(summary)
# 3. 0.87

                            OLS Regression Results                            
Dep. Variable:                 target   R-squared:                       0.518
Model:                            OLS   Adj. R-squared:                  0.507
Method:                 Least Squares   F-statistic:                     46.27
Date:                Thu, 20 Jun 2024   Prob (F-statistic):           3.83e-62
Time:                        01:48:37   Log-Likelihood:                -2386.0
No. Observations:                 442   AIC:                             4794.
Df Residuals:                     431   BIC:                             4839.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const        152.1335      2.576     59.061      0.0